src/AutoRest.CSharp/LowLevel/Output/OperationConfidentChecker.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using AutoRest.CSharp.Common.Input;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace AutoRest.CSharp.Output.Models
+{
+    internal static class OperationConfidentChecker
+    {
+        private static ConcurrentDictionary<InputType, bool> _cache = new();
+
+        public static bool IsConfident(InputOperation operation)
+        {
+            // check parameters
+            foreach (var parameter in operation.Parameters)
+            {
+                // skip the special parameters since some particular types are still interpreted as unions (for instance content type could have union of literal types which is actually enums)
+                if (parameter.IsContentType || parameter.IsApiVersion || parameter.IsEndpoint)
+                    continue;
+
+                var isConfident = IsConfident(parameter.Type);
+                if (!isConfident)
+                    return false;
+            }
+            // check response
+            foreach (var response in operation.Responses)
+            {
+                var isConfident = IsConfident(response.BodyType);
+                if (!isConfident)
+                    return false;
+            }
+
+            // if no low confident encountered on the way, this is confident, we return true.
+            return true;
+        }
+
+        private static bool IsConfident(InputType? type)
+        {
+            if (type == null)
+                return true;
+
+            if (_cache.TryGetValue(type, out var result))
+                return result;
+
+            var visitedModels = new Dictionary<InputModelType, bool?>();
```

Comment:
Why not `HashSet<InputModelType>`?

Reply:
This is not only for "visited models", it is also storages the result of those visited models.
We cannot get the result for one type and just put it into the global cache because sometimes non-confident thing pollutes the entire tree.
I write a test case for this
```
@discriminator("kind")
model Pet {}

model Cat extends Pet { kind: "cat"; }

model Dog extends Pet { kind: "dog"; unionProperty: string | int32; } // <- this type pollutes the entire tree and we need to backport Cat and Pet to be not confident as well
```

Reply:
I see. Another comment. It looks like `IsConfident()` could be invoked many times, and some calculation is duplicated. Maybe we should change `visitedModels` as a static one?

Reply:
No. the idea here is that the global dictionary `_cache` is the final result cache. Things go into that will never be calculated again.
`visitedModels` here is a local cache to cache everything being calculated or calculated in this call.
I do not combine this local cache with global cache because the result of one model is not finalized until the tree has been go over.

I will add a read cache at the beginning of `WalkType` so that if the type has been calculated in previous calls of `IsConfident`, we will not call it again.